File: aizu/tyama_aizuALDS1~7C.rb
#!/usr/bin/ruby
require 'bio'
n=gets.to_i
$tree=Bio::Tree.new
$a=n.times.map{|i|Bio::Tree::Node.new(i)}
$a.each{|e|$tree.add_node(e)}
roots=n.times.to_a
n.times{
	x=gets.split.map(&:to_i)
	p=x.shift
	next if x==[-1,-1]
	x.each{|e|
		$tree.add_edge($a[p],e!=-1 ? $a[e] : Bio::Tree::Node.new(:dummy))
		roots[e]=nil if e!=-1
	}
}
$tree.root=$a[roots.find{|e|e}]
def preorder(node)
	return if node.name==:dummy
	print " #{node.name}"
	$tree.children(node).each{|e|preorder(e)}
end
def inorder(node)
	return if node.name==:dummy
	children=$tree.children(node)
	inorder(children[0]) if children[0]
	print " #{node.name}"
	inorder(children[1]) if children[1]
end
def postorder(node)
	return if node.name==:dummy
	$tree.children(node).each{|e|postorder(e)}
	print " #{node.name}"
end
puts 'Preorder'
preorder($tree.root);puts
puts 'Inorder'
inorder($tree.root);puts
puts 'Postorder'
postorder($tree.root);puts

File: aizu/tyama_aizuALDS1~7C.cpp
#include <vector>
#include <map>
#include <set>
#include <cstdio>
using namespace std;
map<int,vector<int>>H;
void preorder(int node){
	printf(" %d",node);
	for(auto &e:H[node])if(e!=-1)preorder(e);
}
void inorder(int node){
	auto &children=H[node];
	if(children.size()>=1&&children[0]!=-1)inorder(children[0]);
	printf(" %d",node);
	if(children.size()>=2&&children[1]!=-1)inorder(children[1]);
}
void postorder(int node){
	for(auto &e:H[node])if(e!=-1)postorder(e);
	printf(" %d",node);
}
int main(){
	int N,a,b,c;
	scanf("%d",&N);
	set<int>roots;
	for(int i=0;i<N;i++)roots.insert(i);
	for(int i=0;i<N;i++){
		scanf("%d%d%d",&a,&b,&c);
		H[a].push_back(b),roots.erase(b);
		H[a].push_back(c),roots.erase(c);
	}
	int root=*roots.begin();
	puts("Preorder");
	preorder(root);puts("");
	puts("Inorder");
	inorder(root);puts("");
	puts("Postorder");
	postorder(root);puts("");
}
